packages/dss/src/DomainGuest.sol

```diff
@@ -0,0 +1,179 @@
+    event Rectify();
+    event Cage();
+
+    modifier auth {
+        require(wards[msg.sender] == 1, "DomainGuest/not-authorized");
+        _;
+    }
+
+    constructor(address _daiJoin) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        daiJoin = DaiJoinLike(_daiJoin);
+        vat = daiJoin.vat();
+        dai = daiJoin.dai();
+    }
+
+    // --- Math ---
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    function _divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = (x + y - 1) / y;
+    }
+
+    // --- Administration ---
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+
+    function file(bytes32 what, address data) external auth {
+        if (what == "end") end = data;
+        else revert("DomainGuest/file-unrecognized-param");
+        emit File(what, data);
+    }
+
+    /// @notice Set the global debt ceiling for the local dss
+    /// @dev Should only be triggered from the DomainHost
+    function lift(uint256 line, uint256 minted) external auth {
+        vat.file("Line", line);
+        grain += minted;
+
+        emit Lift(line, minted);
+    }
+
+    /// @notice Will release remote DAI from the escrow when it is safe to do so
+    /// @dev Should be run by keeper on a regular schedule
+    function release() external {
+        uint256 limit = _min(vat.Line() / RAY, _divup(vat.debt(), RAY));
+        require(grain > limit, "DomainGuest/no-extra-to-release");
+        uint256 burned = grain - limit;
+        grain = limit;
+
+        _release(burned);
+
+        emit Release();
+    }
+
+    /// @notice Push surplus (or deficit) to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function push() external {
+        // TODO do we want some minimum value to trigger this to give a local buffer?
+        uint256 _dai = vat.dai(address(this));
+        uint256 _sin = vat.sin(address(this));
+        if (_dai > _sin) {
+            // We have a surplus
+            vat.heal(_sin);
+
+            uint256 wad = (_dai - _sin) / RAY;    // Leave the dust
+            daiJoin.exit(address(this), wad);
+
+            // Send ERC20 DAI to the remote DomainHost
+            _surplus(wad);
+        } else if (_dai < _sin) {
+            // We have a deficit
+            vat.heal(_dai);
+
+            _deficit(_divup(_dai - _sin, RAY));   // Round up to overcharge for deficit
+        }
+
+        emit Push();
+    }
+
+    /// @notice Merge ERC20 DAI into surplus
+    function rectify() external {
+        daiJoin.join(address(this), dai.balanceOf(address(this)));
+
+        emit Rectify();
+    }
+
+    function cage() external {
+        end.cage();
+
+        // TODO need to relay the cure value after some settlement period
```

Comment:
Yes, my thoughts is it will essentially match L1 `end` structure. I really want to discuss first if we want to continue supporting GS. It's going to be most of the work here I think.